src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
BTW, I missed the comment of IGrainRuntime having a service locator. I think you are right, we should have an IGrainRuntime scoped service or get rid of most of it, since it seems to be a facade for getting to other services, but in theory you can inject them via DI too. This interface is currently a little odd to as it you could in theory deactivate another grain with it, so it's dangerous to have it be public.